Merge PR Branch to Migration Branch

noti0na1 opened the pull request:
The new flag and NotNull list are rewritten in new style.

I made one change to the behavior of the NotNull list. I will comment it below.

We should merge #35 first and update this branch.

nicolasstucki committed:
Use extension method in TASTy Refection

odersky committed:
Suspend callers of macros compiled in the same run

Allows macros to be defined and called in the same project. The idea is to suspend
the callers of a macro (and their upstream dependencies), compile the sourcefile(s)
containing the macro(s), and then compile the suspended files in a second run.

The scheme detects and flags as errors calls to macros in the same file and mutual
dependcies between different files which would not work with suspensions.

It needs to be complements by another mechanism that detects that class files loaded
by a macro class loader do not have assiciated source files that are currently compiled.

odersky committed:
Make macroDependencies compilation-order independent

Using a Set risks giving different results in different runs.

odersky committed:
Track splice calls instead of inline calls

The real thing that needs to be tracked for cyclic dependencies
are calls to spliced methods. Move the suspend machinery from
inline calls to these.

odersky committed:
Fix tests

Drop reflect-inline, since it did not do what one thought (called stdlib stripMargin
instead of locally defined one).

Move macros-in-same-project1 to run-macros

Fix error marker for macro-cycle1.scala.

odersky committed:
Add `outermost` utility method to SourcePosition.

Not needed right now, but might be useful in the future.

odersky committed:
Add debug output to track down problem compiling ScalaTest

odersky committed:
Add -Xprint-suspension option to print info about suspensions

odersky committed:
Compile scalatest with -Xprint-suspension

odersky committed:
Don't count vals as macro dependencies

nicolasstucki committed:
Move tests to pos-macros

nicolasstucki committed:
Fix error message

nicolasstucki committed:
Fix reflect-inline test

nicolasstucki committed:
Revert "Don't count vals as macro dependencies"

This reverts commit 8cea72067fb5736285a6dbc8a209bddae6ee1e18.

nicolasstucki committed:
Add minimization of previous scalatest stripMargin issue

nicolasstucki committed:
Suspend macros when class not found during macro expansion

nicolasstucki committed:
Improve error message

odersky committed:
Don't suspend if errors were reported.

nicolasstucki committed:
Identify classes defined in current run

nicolasstucki committed:
Remove outdated comment

nicolasstucki committed:
Handle suspended units in InteractiveDriver

nicolasstucki committed:
Test for legitimate NoClassDefFoundError in Splicer

nicolasstucki committed:
Add regression tests

nicolasstucki committed:
Add output as macro classpath for suspended compilation

nicolasstucki committed:
Add error when suspension has jar output

To avoid a compiler crash.

noti0na1 committed:
modify methods in NullOpsDecorator

noti0na1 committed:
rename and adjust methods

noti0na1 committed:
update nullnull test

nicolasstucki committed:
Add supspension to macro docs

OlivierBlanvillain committed:
Fix #6570: Don't reduce MT with empty scrutinies

OlivierBlanvillain committed:
Rename disjoint to provablyDisjoint

The new name is explicit about the function output in case of
uncertainty.

noti0na1 committed:
inline policies in JavaNullInterop

nicolasstucki committed:
Test incremental compilation with suspension

noti0na1 committed:
simplify the result of normNullableUnion

noti0na1 committed:
remove extra space

noti0na1 committed:
optimize widenUnion

odersky committed:
Revert "Enable significant indentation after `return`"

A `return` is not necessarily followed by anything. So our general principles
indicate that indentation should not be significant after it, since otherwise
an off-by-one-space error could change meaning.

odersky committed:
Flag pattern definitions with empty RHS as errors

Crashed in pickler before.

odersky committed:
Flag pattern definitions with empty RHS as errors

Crashed in pickler before.

odersky committed:
Allow multi-variable pattern declarations

Allow multi-variable pattern declarations of the form

    val p1, ..., pn: T

odersky committed:
Allow multi-variable pattern declarations

Allow multi-variable pattern declarations of the form

    val p1, ..., pn: T

odersky committed:
Refactoring: don't consume final `with` in `constrApps`

odersky committed:
Fix position of  in squants

nicolasstucki committed:
Fix typos and improve description

nicolasstucki committed:
Typos and specify run vs compile time

nicolasstucki committed:
Add test on incremental and suspended compilation

anatoliykmetyuk committed:
Merge pull request #7370 from dotty-staging/check-deferred-patdef

Flag pattern definitions with empty RHS as errors

nicolasstucki committed:
Move showName to scala.internal

nicolasstucki committed:
Fix #7048: PCP heal static path dependent types

noti0na1 committed:
rewrite widenUnion

nicolasstucki committed:
Merge pull request #7353 from neshkeev/enable-jdk7-jdk8-tests-sjs

Enable jdk7 jdk8 tests ScalaJS

Batanick committed:
Replacing empty+erased function check with assert

nicolasstucki committed:
Merge pull request #7385 from Batanick/erased_function_assert

Replacing empty+erased function check with assert

nicolasstucki committed:
Merge pull request #7368 from dotty-staging/fix-#7121

Fix #7121: Disallow quotes in annotations

nicolasstucki committed:
Fix #7387: Avoid comparing NoType with other types

nicolasstucki committed:
Merge pull request #7346 from dotty-staging/spliced-names

Add customizable names for definitions in quotes

noti0na1 committed:
remove extra braces

noti0na1 committed:
further simplify widenUnion

noti0na1 committed:
add comment

noti0na1 committed:
Merge pull request #28 from noti0na1/dotty-explicit-nulls-ops

Optimize Methods in NullOpsDecorator; Optimize widenUnion

noti0na1 committed:
combine two type mapper

odersky committed:
Fix #7392: Interpolate all wildcards in the result prototype of a closure

Make sure we interpolate all wildcards in the prototype before making it
the result type of an anonymous function. This was done previously only for
toplevel wildcards, but not for wildcards nested in some type.

odersky committed:
Fix #7383: Don't include nested classes in class bounds

When computing the class bound in an avoid, we approximate
members by refinements. But if the member itself is a class,
we cannot refine with  Classinfo type.

The current fix is to give up in this case and not refine
anything. We could try to refine with an abstract type, but
in that case we have to be careful with F-bounds.

nicolasstucki committed:
Fix #7395: Handle inlined repeated arguments

odersky committed:
Replace wildcard types by TypeVars in decomposeProto

Approximating wildcard types loses too much information. it's better
to replace them by fresh typevars instead.

odersky committed:
Recursively interpolate bounds in interpolateWildcards

nicolasstucki committed:
Merge pull request #7398 from dotty-staging/fix-#7395

Fix #7395: Handle inlined repeated arguments

smarter committed:
Don't unnecessarily compile the bootstrapped compiler

Previously, running `dotc ...` would compile the bootstrapped compiler
even though it runs the non-bootstrapped compiler.

nicolasstucki committed:
Extract Printers from TASTy reflection

nicolasstucki committed:
Use extension method in TASTy Refection

* Remove redundant extension methods
* Use given extension methods

noti0na1 committed:
remove toJavaNullableUnion

smarter committed:
Merge pull request #7400 from dotty-staging/fix-non-bootstrap

Don't unnecessarily compile the bootstrapped compiler

noti0na1 committed:
add a helper for OrType(tp, JavaNull)

noti0na1 committed:
Merge pull request #29 from noti0na1/dotty-explicit-nulls-small

Inline Nullify Policies in JavaNullInterop; combine two TypeMap

odersky committed:
Address review comments

nicolasstucki committed:
Merge pull request #7399 from dotty-staging/use-extension-methods-in-tasty-reflect

Use extension methods in TASTy Reflect

OlivierBlanvillain committed:
Address review

nicolasstucki committed:
Decouple TASTy Reflect printers

nicolasstucki committed:
Merge pull request #7402 from dotty-staging/decouple-tasty-reflect-printers

Decouple TASTy Reflect printers

nicolasstucki committed:
Disambiguate quoted and tasty Type

This ambiguity can be seen in the following
```scala
import scala.quoted._
def f[T: Type](given qctx: QuoteContext) = {
  import qctx.tasty._
  ...
  summon[Type[T]] // qctx.tasty.Type instead of scala.quoted.Type
  ...
}
```

To avoid the ambiguity we simply rename `scala.quoted.Type` to `TypeTag`.
From the 2 this is the one that conforms the least to the common expectation of
what is a `Type`. We chose `TypeTag` because it encapsulates the same kind
of information that `scala.reflect.TypeTag` did.

OlivierBlanvillain committed:
Merge pull request #7364 from dotty-staging/fix-6570-1

Fix #6570: Don't reduce match types with empty scrutinies

smarter committed:
Merge pull request #7397 from dotty-staging/fix-#7383

Fix #7383: Don't include nested classes in class bounds

odersky committed:
Merge pull request #7396 from dotty-staging/fix-#7392

Fix #7392: Interpolate all wildcards in the resulttype of a closure

odersky committed:
Merge pull request #7363 from dotty-staging/change-indent-with

Tweaks to indent syntax

nicolasstucki committed:
Rename and move files

To avoid losing history

biboudis committed:
Merge pull request #7404 from dotty-staging/disambiguate-quoted-type-and-reflection-type

Disambiguate quoted and tasty Type

Commit:
Revert "Disambiguate quoted and tasty Type"

nicolasstucki committed:
Merge pull request #7408 from lampepfl/revert-7404-disambiguate-quoted-type-and-reflection-type

Revert "Disambiguate quoted and tasty Type"

noti0na1 committed:
null is not allowed to be thrown; small edit to the code style

noti0na1 committed:
change type eq test in isNullSpace

odersky committed:
Merge pull request #7391 from dotty-staging/fix-#7387

Fix #7387: Avoid comparing NoType with other types

odersky committed:
Don't generate setters for private fields in desugar

The reason is that some of these fields might be converted later to private[this]
fields which do not get  setter. Instead, generate setters for private vars
that are not private[this] in phase Getters.

harpresing committed:
Add error message when trait parameter prefix is used in its parent

#1589

Lacaranian committed:
Partial fix for lampepfl#7113: Fix Scala.js codegen for BoxedUnit.TYPE

- `scala.runtime.BoxedUnit.TYPE` is now intercepted by `JSCodeGen` as a
  primitive for `ClassOf(ClassRef("V"))`, which Scala.js recognizes as
  `java.lang.Void`
- Enable the javalib/lang/ClassTest for Scala.js, which contains a test
  that uses `BoxedUnit.TYPE`

nicolasstucki committed:
Merge pull request #7393 from etSingh/master

Add error message when trait parameter prefixes are used for its parent

odersky committed:
Do Variance checking at PostTyper

The plan is that by then we know when a members is private[this].

odersky committed:
Make private term members Local if accessed only from this

Make private term members Local if they are accessed only from
their owner's this-type.

odersky committed:
Make private type members Local if accessed only from this

odersky committed:
Fix check file

odersky committed:
Fix #7416: Don't emit signatures of fields of primitive types

Don't emit signatures for private fields of primitive type

odersky committed:
Fix #7416: Don't emit signatures of fields of primitive types

Don't emit signatures for private fields of primitive type

odersky committed:
Better fix: Exclude in DottyBackendInterface

odersky committed:
Better fix: Exclude in DottyBackendInterface

odersky committed:
Deprecate private[this] and protected[this] in 3.1

Add docs to state that they will be dropped.

odersky committed:
Drop protected[this] from compiler codebase

odersky committed:
Drop private[this] from compiler codebase

bishabosha committed:
Make BigDecimalFromDigits FromDigits.Floating[BigDecimal]

The given instance in FromDigits for BigDecimal was incorrectly ignoring exponents in number literals,
This commit instead allows exponents in BigDecimal literals, and adds a test.

anatoliykmetyuk committed:
Fix #7407: Synthesise a quote context only for macros

milessabin committed:
Bump shapeless version

anatoliykmetyuk committed:
Fix #6739 test

There was an error in that test where you need a QuoteContext
in scope to write quotes. The error did not manifest itself since
a synthetic QuoteContext was generated for all inline methods
from which it was required (disregarding whether they are meant as
macro implementations or not). This behavior was wrong, however,
(see #7407) and hence this test needs fixing.

milessabin committed:
Merge pull request #7423 from milessabin/topic/update-shapeless

Bump shapeless version

odersky committed:
Merge pull request #7417 from dotty-staging/fix-#7416

Fix #7416: Don't emit signatures of fields of primitive types

bishabosha committed:
Merge pull request #7420 from bishabosha/bigdec_fromFloating

Make BigDecimalFromDigits extend FromDigits.Floating

liufengyun committed:
Merge pull request #7422 from dotty-staging/i7407

Fix #7407: Synthesise a quote context only for macros

odersky committed:
Fix #7421: Insert missing <outdent> tokens where a region ends

odersky committed:
Fix neg test

noti0na1 committed:
edit comment

odersky committed:
Fix #7425: Handle nullary/parameterless conflicts in infoMeet

We did handle ():T & (=> T), but not (=> T) & ():T.

odersky committed:
Test case for #7175

odersky committed:
Fix doc comment

odersky committed:
Expand test

biboudis committed:
Fix summonFrom match from the documentation

anatoliykmetyuk committed:
Merge pull request #7431 from dotty-staging/fix-doc-summonFrom

Fix summonFrom match from the documentation

smarter committed:
Update CI image to use a proxy

See
https://github.com/lampepfl/dotty-drone/commit/2c350e6ff3bdac2ffcfc0bbe20421469bc59e5aa

noti0na1 committed:
update comment on flags

noti0na1 committed:
Merge pull request #30 from noti0na1/dotty-explicit-nulls-small

* revert changes to `typedThrow`; the expression of `throw` cannot be `null`
* move `flowFacts` in `Context`
* modify comment in `Flags` about `JavaEnumValue`
* remove extra function in `simpleApply`
* remove extra comments

noti0na1 committed:
Merge remote-tracking branch 'upstream/master' into dotty-explicit-nulls

abeln committed:
Merge pull request #31 from noti0na1/dotty-explicit-nulls

Merge Changes from Dotty Upstream

noti0na1 committed:
revert change to throwMethod; clean up if-else

odersky committed:
Fix #7434: Tighten condition for optimizing synthetic objects in LazyVal

odersky committed:
Put back private modifier in test

anatoliykmetyuk committed:
Drop `forceInline` annotation

It was needed to make Dotty cross-compile to Scala 2
before full bootstrap. Its direct semantics
is about adding the `inline` flag to a symbol
annotated by it. `inline` flag, however, is listed as
one of the flags that are not supposed to be modified
by completion of symbols. Hence, having `forceInline`
annotation is in violation of our own assumptions
about completion. Before full bootstrap, we were not
able to drop this annotation, however, now that we
have the full bootstrap, we can do it.

odersky committed:
Fix handling of private trait objects

odersky committed:
Fix #7428: Don't make enum values JavaStatic

It leads to backend crashes if the enum is defined in a class and does
not seem to be needed anyway.

odersky committed:
Make enum values static at Erasure

If an enum value is static (i.e. part of a globally reachable object),
set the JavaStatic flag at erasure.

noti0na1 committed:
Merge pull request #32 from noti0na1/dotty-explicit-nulls

Revert Change about throwMethod; Clean up if-else

noti0na1 committed:
add notnull detect

liufengyun committed:
Fix check file

odersky committed:
Set JavaStatic flag for enum values at Namer

liufengyun committed:
Fix trailing space in error message

liufengyun committed:
Fix CI: remove trailing spaces in check files

Lacaranian committed:
Partial fix for lampepfl#7113: Fix Scala.js codegen for Explicit Returns

In order to use an explicit return statement in Scala.js IR, a label
must be placed wrapping the whole function body to serve as a jump/break
point for all control structures inside the function. While a fresh return
label had been generated for such methods (in the LocalNameGenerator),
it had not been placed at the top level of its method, and was otherwise
unused. This affected the tree created by the TailRec MiniPhase, as well
as user-written early return statements.

After generating a method's body, if a enclosing return label had been
generated, use it to wrap the method body (just as Scala 2's JS codegen
does) for such breaks.

Enable the tests of ReadersTest.scala of the Scala.js test suite, of
which InputStreamReaderTest used a tail-recursive method that depended
on this fix.

anatoliykmetyuk committed:
Merge pull request #7436 from dotty-staging/fix-#7428

Fix #7428: Don't make enum values JavaStatic

bishabosha committed:
prevent separator after 0, update syntax.md

bishabosha committed:
Merge pull request #7441 from bishabosha/fix_separators

Prevent numeric separator after leading zero

anatoliykmetyuk committed:
Merge pull request #7411 from dotty-staging/drop-local

Infer private[this]

smarter committed:
Fix regression: Run the bootstrapped tests with the correct compiler

In bbfe8a3e856b1c09e5f31de5cf4885dd2289727e I switched javaOptions to
use the jars from dotty-compiler instead of dotty-compiler-bootstrapped,
but that means we never actually put the bootstrapped compiler on the
classpath which is of course wrong. Fixed by making javaOptions just
call `packageAll.value` which will get the correct mapping depending on
whether we're in the non-bootstrapped or bootstrapped project, and
dotty-staging/javaOptions now simply forwards to
dotty-compiler-bootstrapped/javaOptions which means we can get rid of
commonDottyJarClasspathSettings.

Also simplified javaOptions in general.

noti0na1 committed:
fix cyclicref

smarter committed:
Merge pull request #7430 from dotty-staging/fix-#7425

Fix #7425: Handle nullary/parameterless conflicts in infoMeet

noti0na1 committed:
Add Support for Parsing Annotations without Arguments to Java Parser (#6893)

bishabosha committed:
Add dotty-staging/intent to community-build

sjrd committed:
Merge pull request #7439 from Lacaranian/add-readers-test

Partial fix for lampepfl#7113: Fix Scala.js codegen for Explicit Returns

smarter committed:
Fully compile primitive value classes

We used to drop compilation units for primitives after frontend, but
this meant that a standard library compiled by dotty was not usable by
dotty, because the primitive value classes were missing from the
classpath: even though these classes are never loaded at runtime,
they're unpickled at compile-time.

Getting this to work required two changes: silence an error in
refchecks (see comment) and turn off some computations in mixins that
require cls.superClass to exist (primitives have no superclasses after
erasure)

noti0na1 committed:
Merge remote-tracking branch 'upstream/master' into dotty-explicit-nulls-small

smarter committed:
Fix LazyRef#ref error recovery

The `throw` keyword was missing, so the call to `CyclicReference` did
not actually do anything. This means that LazyRef#ref could return null
when typechecking broken code, which could lead to a
NullPointerException (I observed this while compiling some code but
didn't manage to minimize it).

noti0na1 committed:
change NotNull string list to ClassSymbol list

noti0na1 committed:
ensure NotNull works for fields

sjrd committed:
Merge pull request #7415 from Lacaranian/add-sjs-class-test

Partial fix for lampepfl#7113: Fix Scala.js codegen for BoxedUnit.TYPE

smarter committed:
Merge pull request #7433 from dotty-staging/ci-cache

Update CI image to use a proxy

smarter committed:
Merge pull request #7443 from dotty-staging/fix-test-run

Fix regression: Run the bootstrapped tests with the correct compiler

smarter committed:
Revert "Revert "Upgrade to ASM 7.0""

This reverts commit 5a747185826bf2b178389b6f48880eb2e5d252a7 which
itself reverted 0abd0769bf27682e1bd62e9532afdafce2c92dfb.

Since #5936 is not fixed yet, so we need to turn off one test in
semanticdb which uses parens in a class name. This is problematic, but
upgrading to ASM 7.0 still seems worth it since it allows running scalac
and dotty in the same classpath which should make it easier to test the
scalac tasty reader.

smarter committed:
Fix documentation generation

The YAML parser used by dottydoc choked on backticks in a title.

abeln committed:
Merge pull request #34 from noti0na1/dotty-explicit-nulls-small

Merge Changes from Dotty Upstream (including changes to JavaParser)

noti0na1 committed:
Merge remote-tracking branch 'explicit-nulls/dotty-explicit-nulls' into dotty-explicit-nulls

smarter committed:
Properly fix documentation generation

Turns out brackets are also special characters, wrapping the value in a
string deals with that.

noti0na1 committed:
Optimize JavaNullInterop; add tests for Java files; remove dummy notnull annotation

bishabosha committed:
Merge pull request #7444 from bishabosha/add-intent-community-build

Add dotty-staging/intent to community-build

smarter committed:
Merge pull request #7448 from dotty-staging/asm-7-again

Revert "Revert "Upgrade to ASM 7.0""

anatoliykmetyuk committed:
Drop the workaround -Yno-inline flag from semanticdb

smarter committed:
Don't run the Java 11 CI tests on open PRs

Java 11 only failures are very rare, so running the CI tests on Java 11
for every commit pushed to an open PR doesn't seem worth it. Instead,
only run these tests on merged commits and release builds.

anatoliykmetyuk committed:
Merge pull request #7450 from dotty-staging/less-java11

Don't run the Java 11 CI tests on open PRs

noti0na1 committed:
rename variables

noti0na1 committed:
add generic tests

noti0na1 committed:
fix typos and indents

noti0na1 committed:
remove redundent comment

noti0na1 committed:
update comments

odersky committed:
Merge pull request #7449 from dotty-staging/semanticdb-noindent

Drop the workaround -Yno-inline flag from semanticdb

odersky committed:
Merge pull request #7447 from dotty-staging/compile-primitives

Fully compile primitive value classes

odersky committed:
Merge pull request #7435 from dotty-staging/fix-#7434

Fix #7434: Tighten a condition in LazyVals

odersky committed:
Merge pull request #7446 from dotty-staging/fix-lazyref

Fix LazyRef#ref error recovery

odersky committed:
Merge pull request #7386 from dotty-staging/fix-#7048-2

Fix #7048: PCP heal path dependent types

odersky committed:
Fix #6938: Expand tuples when synthesizing given names

Also: Add more explanations on how names are synthesized.

odersky committed:
Syntax change for type parameters of extension methods

The new syntax has the type parameters come first. I.e.
```
  def [T](xs: List[T]) append (ys: List[T]): List[T] = ...
```
instead of
```
  def (xs: List[T]) append [T] (ys: List[T]): List[T] = ...
```
Application is unchanged, so it is still
```
  xs.append[String](ys)
```
An argument for the old syntax is that it aligns definition and call syntax.
On the other hand, the new syntax maintains the general rule that parameter introductions
always com before parameter uses. The decisive argument to switch is to be consistent
with the new collective parameter syntax, where `append` would be written like this:
```
  given [T](xs: List[T])
    def append (ys: List[T]): List[T] = ...
```
To avoid misalignment of type parameters between definition and call syntax, we considered
disallowing explicit type parameters for extension methods altogether, and to require
that the method is called as a normal method instead. But that would not work for anonymous
givens as in the last exaple above.

x3ro committed:
Port 'unknown named enclosing class or object' error to new scheme

Part of the effort documented in #1589 to port all error messages to
the new scheme.

odersky committed:
Fix #7438: Turn failing assertion into error in inliner

solnaranu committed:
Fix typo

milessabin committed:
Merge pull request #7454 from dotty-staging/fix-#7075

Fix #7075: Fix implicit resolution spec

milessabin committed:
Merge pull request #7453 from dotty-staging/fix-#6938

Fix #6938: Expand tuples when synthesizing given names

x3ro committed:
Fix typo in explanation

OlivierBlanvillain committed:
Merge pull request #7457 from x3ro/lucas/unknown-named-class-or-object

Port 'unknown named enclosing class or object' error to new scheme

anatoliykmetyuk committed:
Merge pull request #7324 from dotty-staging/add-suspend

Fix #7138: Suspend callers of macros compiled in the same run

noti0na1 committed:
update comments and var names

noti0na1 committed:
Merge pull request #33 from noti0na1/dotty-explicit-nulls

Detect NotNull Annotations for Java Methods

noti0na1 committed:
merge dotty upstream

odersky committed:
Merge pull request #7429 from dotty-staging/fix-#7421

Fix #7421: Insert missing <outdent> tokens where a region ends

noti0na1 committed:
merge PR branch

noti0na1 commented on compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:
```diff
@@ -57,117 +57,26 @@ object JavaNullInterop {
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
+    else if (ctx.settings.YJavaInteropDontNullifyOutermost.value) 
+      nullifyExceptReturnType(tp)
```
I think we can combine the method case and the field case now.

abeln replied:
Right, I agree.

noti0na1 commented on compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:
```diff
@@ -260,38 +168,50 @@ object JavaNullInterop {
-          NoOpPolicy(_ == sym)
-        case None => FalsePolicy
-      }
+        case Some(_) => true
```
For fields in the list, we should still nullify the inside type without a `| JavaNull` at the outermost, similar to `@NotNulll` annotation on a field.

abeln replied:
Yes, you're right. The previous code was incorrect. Nice catch!

odersky committed:
Add neg test

odersky committed:
Fix docs and Parser

x3ro committed:
Port "illegal cyclic type reference" error to new scheme

Part of the effort documented in #1589 to port all error messages to
the new scheme.

odersky committed:
Fix #7459: Fix two crash conditions in Inliner

odersky committed:
Add unrelated test

I wanted to get the test in anyway and this will hopefully get the CI unstuck.

odersky committed:
Fix neg test

odersky committed:
Fix neg test

abeln committed:
Merge pull request #35 from noti0na1/dotty-explicit-nulls

Merge Changes from Dotty Upstream

abeln commented:
#35 is now merged

abeln commented on compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala:
```diff
@@ -57,117 +57,26 @@ object JavaNullInterop {
+      nullifyExceptReturnType(tp)
+    else if (ctx.settings.YJavaInteropDontNullifyOutermost.value) 
+      nullifyExceptReturnType(tp)
+    else if (isInNotNullStdLibList(sym, tp))
```
We can merge this branch with the previous one.

abeln reviewed (approved):
This looks good. Before merging, could you please test it by re-running the community build and checking that the errors counts match what we had before?

martijnhoekstra committed:
Indicate OS support in requirements

odersky committed:
Merge pull request #7455 from dotty-staging/change-extmethod-tparams

Syntax change for type parameters of extension methods

biboudis committed:
Merge pull request #7470 from solnaranu/docs-multiversaleq-typo

Fix multiversal equality docs typo

odersky committed:
Introduce `open` modifier on classes

noti0na1 committed:
Merge remote-tracking branch 'explicit-nulls/dotty-explicit-nulls' into dotty-explicit-nulls-migration

noti0na1 committed:
combine cases

noti0na1 commented:
@abeln I select several projects to test (scalacheck, scalatest, betterfile). The error counts are all different. 
Error counts (checker framework metadata) are slightly higher (except scalatest which has a lot more errors now 46 -> 226). 
The Error counts (java-interop-dont-nullify-outermost) are slightly lower.
I also tested the new project: intent, it has 10 and 5 errors without and with the flag respectively.

odersky committed:
Treat non-open classes as effectively sealed.

This does not affect pattern matching exhaustivity (yet) since currently
only abstract sealed classes and sealed traits are checked for exhaustivity,
so defualt classes do not count. I wonder whether we should change that (?).

odersky committed:
Fix `isInaccessibleChildOf`

nicolasstucki committed:
Merge pull request #7461 from dotty-staging/fix-#7438

Fix #7438: Turn failing assertion into error in inliner

nicolasstucki committed:
Extract modules in reflection Type

This was done on Tree an all other types in TASTy reflect but we forgot to do it for Type

odersky committed:
Set JavDefined flag for Object

Also, make all synthetic classes produced in Definitions Open.

abeln commented:
@noti0na1 thanks for checking

Could you please do the following
1) Pick any checker framework project (preferably one with fewer errors)
2) Run it with the branch before this PR and record the errors
3) Run it with the branch after this PR and record the errors
4) As per your message, there should be more errors in 3) than in 2)

What we want to understand is whether there are more errors because a problem in the merge, or because a problem in the _previous_ branch (which we already suspect to be the case because of the error you fix in JavaNullInterop).

If you could look into that, that would be great. If you could send me the error logs for 2) and 3), I can take a look as well.

Thanks!

cc @olhotak

nicolasstucki committed:
Merge pull request #7469 from martijnhoekstra/patch-1

Indicate OS support in requirements

nicolasstucki committed:
Move extension methods near their modules

nicolasstucki committed:
Fix #7189: Do not try to load contents if file does not exist

anatoliykmetyuk committed:
Merge pull request #7475 from dotty-staging/remove-forceinline

Drop `forceInline` annotation

nicolasstucki committed:
Merge pull request #7473 from dotty-staging/flatten-tasty-reflect-type

Extract modules in reflection Type

odersky committed:
Require language import for ad hoc extensions

noti0na1 commented:
@abeln I just found another potential bug in `JavaNullMap` of `JavaNullInterop`.

```scala
def needsNullArgs(tp: AppliedType): Boolean = !tp.classSymbol.is(JavaDefined)

// map
case appTp @ AppliedType(tycons, targs) =>
  val oldOutermostNullable = outermostLevelAlreadyNullable
  outermostLevelAlreadyNullable = false
  val targs2 = if (needsNullArgs(appTp)) targs map this else targs
  outermostLevelAlreadyNullable = oldOutermostNullable
  val appTp2 = derivedAppliedType(appTp, tycons, targs2)
  if (needsNull(tycons)) toJavaNullableUnion(appTp2) else appTp2
```

When nullify `List[Array[String]]`, the result would be `List[Array[String]]`, since `List` is Java Defined and `needsNullArgs(appTp) == false`. But I think a correct answer should be `List[Array[String | JavaNull]]`.

A fix:

```scala
case appTp @ AppliedType(tycons, targs) =>
  val oldOutermostNullable = outermostLevelAlreadyNullable
  // change this two lines
  outermostLevelAlreadyNullable = tp.classSymbol.is(JavaDefined)
  val targs2 = targs map this
  outermostLevelAlreadyNullable = oldOutermostNullable
  val appTp2 = derivedAppliedType(appTp, tycons, targs2)
  if (needsNull(tycons)) toJavaNullableUnion(appTp2) else appTp2
```

abeln commented:
@noti0na1 good find!
Yes, I agree that we have a bug, and I agree with your fix. Could you please submit that fix as a separate PR to the explicit nulls branch. Could you please add your test as well as a few more
that exercise increased nestedness levels?

e.g.

```Scala
List[Array[List[Array[String]]]] => List[Array[List[Array[String|JavaNull]]|JavaNull]]|JavaNull
```

Thanks!

nicolasstucki committed:
Merge pull request #7476 from dotty-staging/fix-#7189

Fix #7189: Do not try to load contents if file does not exist

odersky committed:
Add doc page

odersky committed:
Merge pull request #7466 from dotty-staging/fix-#7459

Fix #7459: Fix two crash conditions in Inliner

odersky committed:
Don't treat default classes as sum types

Classes representing sum types need to be be abstract, which default classes are not.

This partially reverts "Treat non-open classes as effectively sealed."
d98c2d62defcda679a03a34ee16dd4ef2d103514.

odersky committed:
Fixes to doc page

anatoliykmetyuk committed:
Add 20th release blog article

anatoliykmetyuk committed:
Merge remote-tracking branch 'origin/0.19.x'

anatoliykmetyuk committed:
Set baseVersion to 0.21.0

odersky committed:
Another fix for isInaccessibleChildOf

odersky committed:
Number list items in doc page

noti0na1 committed:
fix nullifying type arguements

noti0na1 committed:
fix indent

liufengyun committed:
Merge pull request #7465 from x3ro/lucas/illegal-cyclic-type-reference

Port "illegal cyclic type reference" error to new scheme

anatoliykmetyuk committed:
Upgrade Dotty to 0.20.0-RC1

nicolasstucki committed:
Merge pull request #7481 from dotty-staging/dotty-release-0.20.0-RC1

Upgrade Dotty Reference to 0.20.0-RC1

martijnhoekstra committed:
allow $ quote escape

odersky committed:
Merge pull request #7486 from martijnhoekstra/sipunquote

escape quote characters in interpolations with $

odersky committed:
Merge pull request #7471 from dotty-staging/add-open-class

Introduce `open` modifier on classes

Commit:
Don't move worksheet decoration when adding new line

Previously, when using worksheets, the evaluated code would move to the
next line when adding a new line. Now, the evaluated code stays on the
original line.

smarter committed:
Merge pull request #7489 from olafurpg/decoration-range-behavior

Don't move worksheet decoration when adding new line

abeln commented:
This will need to be updated after https://github.com/abeln/dotty/pull/37 is merged, right?

nicolasstucki committed:
Fix #7488 and fix #7487: Get String form ConstantType

anatoliykmetyuk committed:
Merge pull request #7480 from dotty-staging/20-release-blog

20th Release Blog Article

nicolasstucki committed:
Merge pull request #7490 from dotty-staging/fix-#7488

Fix #7488 and fix #7487: Get String form ConstantType

noti0na1 committed:
fix typo in the test

noti0na1 committed:
Merge pull request #37 from noti0na1/dotty-explicit-nulls

Fix nullification of type arguments

noti0na1 committed:
merge upstream changes

noti0na1 committed:
fix syntax error

abeln committed:
Merge pull request #38 from noti0na1/dotty-explicit-nulls

Merge Changes from Dotty Upstream

noti0na1 committed:
Merge remote-tracking branch 'explicit-nulls/dotty-explicit-nulls' into dotty-explicit-nulls-migration

noti0na1 commented:
@abeln updated

abeln commented:
Did you do a full test run for this? Do all tests pass?

noti0na1 commented:
@abeln Yes, all tests pass

abeln commented:
Great! Thanks for doing this. When you have a chance, could you please do a full run of the community build and update the spreadsheet? We're mainly interested in the errors counts _with_ the don't-nullify-java-types flag enabled.
